desktop_version/src/Font.cpp

```diff
@@ -0,0 +1,345 @@
+        {
+            f->glyph_h = help.Int(pElem->GetText());
+        }
+        if ((pElem = hDoc.FirstChildElement("white_teeth").ToElement()) != NULL)
+        {
+            // If 1, we don't need to whiten the entire font (like in old versions)
+            white_teeth = help.Int(pElem->GetText());
+        }
+    }
+
+    f->image = LoadImage(name_png, white_teeth ? TEX_COLOR : TEX_WHITE);
+    SDL_zeroa(f->glyph_page);
+
+    if (f->image == NULL)
+    {
+        return;
+    }
+
+    /* We may have a 2.3-style font.txt with all the characters.
+     * font.txt takes priority over <chars> in the XML.
+     * If neither exist, it's just ASCII. */
+    bool charset_loaded = false;
+    unsigned char* charmap = NULL;
+    size_t length;
+    if (FILESYSTEM_areAssetsInSameRealDir(name_png, name_txt))
+    {
+        FILESYSTEM_loadAssetToMemory(name_txt, &charmap, &length, false);
+    }
+    if (charmap != NULL)
+    {
+        charset_loaded = true;
+        unsigned char* current = charmap;
+        unsigned char* end = charmap + length;
+        int pos = 0;
+        while (current != end)
+        {
+            uint32_t codepoint = utf8::unchecked::next(current);
+            add_glyphinfo(f, codepoint, pos);
+            ++pos;
+        }
+
+        VVV_free(charmap);
+    }
+
+    if (xml_loaded)
+    {
+        if (!charset_loaded && (pElem = hDoc.FirstChildElement("chars").ToElement()) != NULL)
+        {
+            // <chars> in the XML is only looked at if we haven't already seen font.txt.
+            int pos = 0;
+            tinyxml2::XMLElement* subElem;
+            FOR_EACH_XML_SUB_ELEMENT(pElem, subElem)
+            {
+                EXPECT_ELEM(subElem, "range");
+
+                unsigned start, end;
+                if (!decode_xml_range(subElem, &start, &end))
+                {
+                    continue;
+                }
+
+                for (uint32_t codepoint = start; codepoint <= end; codepoint++)
+                {
+                    add_glyphinfo(f, codepoint, pos);
+                    ++pos;
+                }
+            }
+            charset_loaded = true;
+        }
+
+        if ((pElem = hDoc.FirstChildElement("special").ToElement()) != NULL)
+        {
+            tinyxml2::XMLElement* subElem;
+            FOR_EACH_XML_SUB_ELEMENT(pElem, subElem)
+            {
+                EXPECT_ELEM(subElem, "range");
+
+                unsigned start, end;
+                if (!decode_xml_range(subElem, &start, &end))
+                {
+                    continue;
+                }
+
+                int advance = subElem->IntAttribute("advance", -1);
+                int color = subElem->IntAttribute("color", -1);
+
+                for (uint32_t codepoint = start; codepoint <= end; codepoint++)
+                {
+                    GlyphInfo* glyph = get_glyphinfo(f, codepoint);
+                    if (glyph != NULL)
```

Comment:
Is it okay if I only fix this in the latest commit which inserted that code? This is a case where the code moved around and evolved a whole bunch, and I'd have to dig deep to find every commit which individually places/moves this and repeat myself while either fixing conflicts or carefully peeling commits apart. (Or should I squash everything?)

Reply:
Okay sure, it can be fixed only in the latest commit.

Reply:
Ok, actually, I found a way to make repeating fixes across commits more bearable: if I `git rebase -i -Xtheirs upstream/master`, the only part I have to worry about is repeating myself across every commit - not also about fixing all the conflicts it causes (since it'll just ignore the fix made to the earlier commit when applying the more recent commit). I'd still recommend if you find a problem during review to also check the latest commit to see if it's still there, but this should at least make the process less of a time-waster for both of us.